충남대 김규현 Step1 카카오 로그인

HappyGogildong opened the pull request:
저는 **RestClient**를 사용했습니다.
Spring에서 요청을 보내는 방법은 RestTemplate, RestClient, WebClient, Http Interface 4가지 방법을 찾아보았습니다
WebClient는 Webflux 의존성을 추가해야 하고 reactive 시스템에 사용하는 방법이라고 되어있어 이 프로젝트와는 맞지 않는다고 생각했습니다.
RestClient가 RestTemplate을 대체하는 기술이라 사용했습니다.
나중에 카카오 API로 보내는 요청이 다양해지면 HttpInterface를 도입하는 게 적합하다고 생각했습니다.

HappyGogildong committed:
<build>(project): 0단계 코드 준비

HappyGogildong committed:
<feat>(kakaoAuth): 카카오톡 권한이 들어있는 토큰 발급받는 기능 구현

HappyGogildong committed:
<feat>(kakaoAuth): 권한 인증된 후 리다이렉트 된 URL에서 authKey를 바로 가져오게 변경

HappyGogildong committed:
Merge branch 'happygogildong' into step1

seungy0 commented on src/main/java/gift/controller/KakaoController.java:
```diff
@@ -0,0 +1,30 @@
+
+    @GetMapping("/")
+    public void handleAuthKey(@RequestParam Map<String, String> params) {
+        authKey = params.get("code");
```
혹시 어떤 연유로 Controller에 멤버 변수로 authKey 변수를 선언하고 여기에 값을 담아서 사용하기로 결정하셨는지 궁금합니다.
이 경우 동시성 문제가 발생할 수 있습니다.

authKey를 저장해두고 사용하는 방식은 사용자가 1명일 때만 유효한 방식이라서 조금 더 고민해보셔도 좋을 것 같아요!

궁금하신 부분 있으면 질문주세요~

HappyGogildong replied:
하지만 지금과 같이 클라이언트가 없는 상황에서는 아래와 같은 방법으로 인가 코드를 획득한다.

내 애플리케이션 > 앱 설정 > 앱 키로 이동하여 REST API 키를 복사한다.
https://kauth.kakao.com/oauth/authorize?scope=talk_message&response_type=code&redirect_uri=http://localhost:8080&client_id={REST_API_KEY}에 접속하여 카카오톡 메시지 전송에 동의한다.
http://localhost:8080/?code={AUTHORIZATION_CODE}에서 인가 코드를 추출한다.

과제 페이지에서 인가 코드를 얻어오는 방법인데, 한 번 사용한 인가 코드는 다시 토큰 발급에 사용하려고 하면 KOE320 에러가 발생했습니다. 그래서 테스트 할 때 인가 코드를 계속 URL에서 복붙했는데, 그래서 redirect url을 매핑해서 복붙하지 않고 바로 인텔리제이에서 확인하고 토큰 발급까지 하려고 임시로 만들었습니다.

등록한 redirect url에서 토큰 발급까지 하게 변경하였습니다

HappyGogildong replied:
저는 http://localhost:8080/kakao-auth 
url을 사용해서 토큰 발급까지 같은 url에서 하도록 만들어놨습니다
https://kauth.kakao.com/oauth/authorize?scope=talk_message&response_type=code&redirect_uri=http://localhost:8080/kakao-auth&client_id={REST_API_KEY}]

seungy0 commented on src/main/java/gift/kakao/KakaoAuthService.java:
```diff
@@ -0,0 +1,47 @@
+                .retrieve()
+                .toEntity(String.class);
+        } catch (Exception e) {
+            System.out.println(e);
```
로거를 사용하시면 좋을 것 같습니다~
스프링에서 기본적으로 제공하는 `@Slf4j` 를 사용하시면 쉽게 구현가능합니다.
로거를 사용하는게 어떤 차이가 있을까요?

HappyGogildong replied:
@Slf4j 는 Lombok을 추가해야되서 사용하지 못 할 것 같습니다..

https://developers.kakao.com/docs/latest/ko/kakaologin/trouble-shooting#token
https://cloud.google.com/apigee/docs/api-platform/reference/policies/oauthv2-policy?hl=ko#errorreference
https://datatracker.ietf.org/doc/html/rfc6749#section-5.2

구글의 링크를 보면 상태 코드 별로 다양한 메세지가 있어서, 일단 exception을 받고,
그 상태 코드와 바디를 그냥 Body에 담아 반환하게 했습니다. 어떤 오류가 올 지 몰라서 이렇게 구현했습니다.
그리고 제 서버의 문제와 카카오 인증 서버의 문제를 구분하기 위해 두 catch를 활용했습니다.

그리고 exception 디렉토리가 이젠 좀 복잡해진 것 같아서 하위 디렉토리에 분류를 하고자 하는데 멘토님은 어떻게 생각하시나요?

seungy0 commented on src/main/java/gift/kakao/KakaoAuthService.java:
```diff
@@ -0,0 +1,47 @@
+
+    public ResponseEntity<String> getAuthorization(String authKey) {
+        ResponseEntity<String> response = null;
+        RestClient restClient = RestClient.create();
```
RestClient 사용 👍👍
매번 RestClient를 생성하고 있습니다.
빈 주입을 추천드려요.
또는 서비스 클래스 생성자에서 초기화한 뒤에 사용하는 것도 가능하긴 합니다ㅎㅎ

HappyGogildong replied:
빈 주입을 관성적으로 클래스 파일에만 사용하고 있었는데, 객체라면 다 사용가능한건가요? 

저는 RestClient를 사용하기 위해 공식 문서, 사람들의 블로그를 찾아봤었는데 RestClient 구현하기, Spring에서 Http 요청 보내기 등으로 검색했었습니다.
멘토님이 피드백 주신 것처럼 어떤 기술을 더 잘 활용하는 방법 같은 건 혼자서 어떻게 찾아봐야할까요?

구현이 아닌 더 나은 활용방안을 찾는 방법이 궁금합니다

seungy0 commented on src/main/java/gift/kakao/KakaoAuthService.java:
```diff
@@ -0,0 +1,47 @@
+        ResponseEntity<String> response = null;
+        RestClient restClient = RestClient.create();
+
+        String url = "https://kauth.kakao.com/oauth/token";
```
환경변수로 빼는 편이 좋겠습니다.

HappyGogildong replied:
application.properties에 넣어서 관리하도록 변경했습니다.

seungy0 commented on src/main/java/gift/kakao/KakaoAuthService.java:
```diff
@@ -0,0 +1,47 @@
+    private String redirectUri;
+
+
+    public ResponseEntity<String> getAuthorization(String authKey) {
```
내부 로직이 액세스 토큰 발급이라 메서드명과 일치하지 않는 부분이 있습니다.

HappyGogildong replied:
getAuthToken으로 변경했습니다

seungy0 commented on src/main/java/gift/controller/KakaoController.java:
HttpInterface 사용여부와 별개로 외부 API 호출하는 부분 서비스 클래스에서 분리하는 편이 유지보수에 유리합니다.

이 때 HttpInterface를 사용하시면 좀 더 편하겠죠ㅎㅎ

HappyGogildong replied:
RestClient도 비슷한 맥락에서 분리하여 빈 주입을 추천하신거죠?
다음 미션에서 HttpInterface를 활용해 보겠습니다

seungy0 commented:
규현님 안녕하세요.
외부 API 호출을 어떻게 할 것인지에 대해 고민하신 부분 잘봤습니다 👍👍
몇가지 코멘트 남겼으니 참고해주세요~

seungy0 reviewed (changes requested):
규현님 안녕하세요.
외부 API 호출을 어떻게 할 것인지에 대해 고민하신 부분 잘봤습니다 👍👍
authKey 관련 문제는 이번에 수정하고 넘어가면 좋을 것 같아서 변경요청드리겠습니다.
추가로 몇가지 코멘트 남겼으니 참고해주세요~

HappyGogildong committed:
rebase

HappyGogildong committed:
<refactor>(kakaoAuthService): 토큰 발급 에러 시 처리 로직 변경
카카오 인증 서버 url 환경변수(application.properties)에서 관리
RestClient를 @Configuration로 등록하여 사용

HappyGogildong committed:
<refactor>(KakaoAuthService):getAuthorization -> getAuthToken으로 메서드명 변경

HappyGogildong committed:
<fix>(환경 변수):redirect uri가 변경됨에 따라 환경 변수 수정

HappyGogildong committed:
<refactor>(토큰 Dto):카카오 Auth token 반환을 String에서 Dto로 바꿈

HappyGogildong commented:
push를 하려면 rebase를 먼저 하라고 해서 했는데 제 기존 커밋에 되는 줄 모르고 제목을 바꿔버렸습니다
rebase 커밋의 기존 이름은 
 <refactor> (kakaoController) : 리다이렉트 uri에서 토큰 발급도 같이 하게 변경
입니다.

seungy0 commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -133,6 +143,9 @@ public String handleKakaoApproveException(
             e.getMessage());
         return "redirect:/managerHome";
     }
+
+
+
```
nit: 불필요한 개행입니다.

seungy0 commented on src/main/java/gift/dto/response/KakaoAuthTokenResponse.java:
```diff
@@ -0,0 +1,12 @@
+package gift.dto.response;
+
+public record KakaoAuthTokenResponse(
```
응답 record class 👍
다만 snake case를 변수명으로 그대로 사용하지 않고 JsonProperty와 같은 기능을 사용해서 변수명을 카멜 케이스로 통일하면 좋겠습니다.

seungy0 commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -125,6 +125,16 @@ public ResponseEntity<ExceptionResponseDto> handleOptionNotFoundException(
+    ) {
+        ExceptionResponseDto exception = ExceptionResponseDto.singleIssue(e.getMessage(),
+            LocalDateTime.now());
+        return exception;
```
200으로 응답하는 것을 의도하신 것일까요~?

seungy0 reviewed (approved):
규현님 고생하셨습니다~
/oauth/authorize는 FE에서 호출한다고 가정하신 것으로 보입니다. 맞나요?ㅎㅎ
그리고 코드를 살펴보던 중에, `AuthorizationInterceptor` 에 헤더명을 Authorization이 아니라 Bearer로 잘못 사용하고 계신 부분이 있더라구요.
이 부분도 확인 부탁드립니다.
